app/controllers/application_controller.rb

```diff
@@ -1,4 +1,22 @@
 class ApplicationController < ActionController::Base
   # TODO: Remove when devise is added
-  attr_reader :current_user
+  def user_logged_in?
+    !!current_user
+  end
+
+  def current_user
+    User.first
+  end
+
+  %w[index show].each do |action|
+    define_method action do
+      if user_logged_in? && respond_to?("authenticated_#{action}")
+        send("authenticated_#{action}")
+        render action: "authenticated/#{action}"
+      elsif !user_logged_in? && respond_to?("external_#{action}")
+        send("external_#{action}")
+        render action: "external/#{action}"
+      end
```

Comment:
Ruby meta-programming in action!

Reply:
Will this be removed once we add Devise in? If not, I'd rather have this hard coded since it's easier to read.

Reply:
By hard-coded, what do you mean? Duplicated in every controller?

Reply:
I'm merging this for now but will be refactoring it later.

Reply:
Not in every controller, just defined as separate `index` and `show` actions in `ApplicationController`.

I've been reading the code more and I understand what you want to get out of this, maybe we could write a helper method for this that is called within the `index` and `show` actions rather than defining the actions programatically. It might be just personal preference that I want the crud actions, especially in controllers, as transparent as possible.

Reply:
I chatted quite a bit 1-1 with @SleeplessByte too. He suggested having a DSL for this, which I liked.

So rather than having this global block of code in application_controller, we might instead add the following to any controller where we want to have an external index and show page:

```
external :index, :show
```

Which would then do something like add a `before_action` that checks if the person is logged in, and if not, uses the `external_index` or `external_show` action rather than the `index` action.

So sort of somewhere between what I've got here (full automated) and what you suggest (quite manual).

The main reason for this, btw, is that I want to have consistent URLs across use cases. So someone logged out would go to `/tracks/ruby/exercises/bob` for the SEO page and someone logged in would go to the **same** URL for their solution to Ruby/Bob (whereas at the moment, everything internal gets namespaced to `my/...` and we have things like `/my/solution/xxxx` which is a non-sharable link.

So because you can't pivot the controller based on state in the routes (without writing custom middleware, which **might** be an option, but isn't a rabbit hole I want to go into now), pivoting the actions in the controller feels like a better option.

Reply:
Great idea!!!!!!

Reply:
Thanks :)